csi driver mounts of host file system content into pod without hostpa… (#250)
openshift/enhancements: enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md

users' workload Pods so that containers in those Pods can utilize those CAs when performing TLS based communication.

3. See [this proposal](https://github.com/openshift/enhancements/pull/214) for additional user scenarios that could 
utilize this proposal.

### Implementation Details/Notes/Constraints [optional]

#### Install

A comparison of installing the implementation of this proposal via OLM vs. CVO was done.  The net result was 
to go with CVO.

Some specifics:
- During the review for the ["Installation of CSI drivers that replace in-tree drivers](https://github.com/openshift/enhancements/blob/master/enhancements/storage/csi-driver-install.md)
a question was posed about whether the exact type of CSI driver this proposal is describing would use the method 
described in that document (namely OLM).  The recommendation from the [storage SMEs was to go with CVO](https://github.com/openshift/enhancements/pull/139/files#r391761362)
- In particular, using CVO and having it be a day 1 install operation meets the ease of use targets we are trying to
reach with this proposal

> - Crafting dependencies between this proposal and [related proposals](https://github.com/openshift/enhancements/pull/214)

Author: yes ... I'll clarify

their Pod privileged, I want a means for doing that.

3. See [this proposal](https://github.com/openshift/enhancements/pull/214) for additional user scenarios that could 
utilize this proposal.

### Implementation Details/Notes/Constraints [optional]

#### Brief overview of CSI and the main K8S hook points

 - a [CSI plugin](https://github.com/container-storage-interface/spec) is a gRPC endpoint that implements CSI services and
 allows storage vendors to develop a storage solution that works across a number of container orchestration systems.
 - though for our purposes, we will not provide a generic CSI plugin, but rather a special case one that is only intended
 to work on OpenShift
 - also for our purposes, `plugin`, `driver`, and `endpoint` seem to be the same thing based on which document you are
 reading.
 - k8s provides Kubernetes CSI Sidecar Containers: a set of standard containers that aim to simplify the development and 
 deployment of CSI Drivers on Kubernetes.
 - There are CSI volume and persistent volume types in k8s.
 - per [k8s volume docs](https://kubernetes.io/docs/concepts/storage/volumes/#csi) the CSI types do not support direct

>  reference from a Pod and may only be referenced in a Pod via a `PersistentVolumeClaim`

Reviewer: This is not true. That is where inline CSI volumes come - https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html

Author: out of date info at https://kubernetes.io/docs/concepts/storage/volumes/#csi  ... not *necessarily* surprising :-)

thanks for the clarification @gnufied 

I had noticed the inline feature touched upon in a few places, but had not dived into yet.

Aside from updating these notes to reflect the discrepancy in https://kubernetes.io/docs/concepts/storage/volumes/#csi and what inline provides, I'll read up on inline CSI and see if it has any bearing on our scenario.

Of course chime in as you see fit if you already see a place where it might.

thanks again

---
title: inject-cross-namespace-secrets-onto-openshift-hosts-for-general-pod-consumption

authors:
  - "@gabemontero"

reviewers:
  - "@bparees"
  - "@adambkaplan"
  - "@derekwaynecarr"
  - "@gnufied"
  
approvers:
  - "@bparees"
  - "@adambkaplan"
  - "@derekwaynecarr"
  - "@gnufied"
  
creation-date: 2020-03-17

last-updated: 2020-03-18

<!-- status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced -->
status: provisional

  
---

# Inject cross namespace secrets onto OpenShift Hosts for general Pod consumption

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [/] Design details are appropriately documented from clear requirements
- [/] Test plan is defined
- [/] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Open Questions [optional]

> 1. In various conversations leading up to this proposal, a consensus has yet to arrive on how to package and deliver
>a CSI driver/plugin based solution into OpenShift.  Some alternatives include:
> > a) As part of a CVO based operator that is included in a 4.x OCP install

> > > b) An OLM Operator deployed solution

Reviewer 2: since this seems like a feature we expect to work seemlessly for users (when we have automatic subscription someday) and since we don't currently have the ability to auto-install an OLM operator as part of a day 1 install, the CVO approach seems like the only way to go here.  But certainly i'm interested to hear who is pushing for OLM (presumably from a desire to keep from expanding the default cluster install footprint).

doing it via olm also means setting up some sort of dependency between the csi driver and the admission controller (also packaged as OLM?  If both are OLM, olm does have a concept of dependent operators, but the api has a lot of limitations/undefined behavior right now)

Author: @adambkaplan may have gotten it from somewhere else, but I got the notion from #139 

However, our scenario here came up during the review, after my initial discovery of it, during an exchange between @deads2k and @jsafrane ,  and "use CVO, not OLM" was the apparent answer as well:  https://github.com/openshift/enhancements/pull/139/files#r391761362

I'll given @adambkaplan a few cycles to chime in, but otherwise, I'll put a stake in the ground wrt CVO

### Non-Goals

1. This proposal assumes the sensitive content encapsulated in the K8S Secret is valid and correct for the user's purposes.
No validation of the content for specific usage within a Pod will be done.  That said, this proposal does provide 
for watching for updates to the K8S Secrets being stored on the host/node by the associated CSI Driver of this proposal.
So if the cluster administrator realized on their own they have provided bad data, and then corrects it, this proposal
will watch for, discover, and propagate the update to the host/node filesystem.

2. Protecting this content from being viewed/copied by users. If it's available for the Pod to use, the user who can 
create a Pod can also read / view / exfiltrate the Secret contents. It is technically impossible to avoid that.

3. 4.x customer feedback has also revealed a desire for namespace level granularity when injecting/consuming
K8S Secret information available on the underlying host/node.  While this proposal will articulate some thought on 
how to do this, ultimately a solution based on a given Secret utilized across multiple namespaces is deemed "meets

> minimum" for this proposal, and per namespace secret injection is out of scope.   

Reviewer 2: arguably this proposal(specific to the CSI driver) doesn't concern itself with whether the content is "global" or user namespaced.

in implementation terms, there's no difference.  the CSI driver is going to pull content from some namespace (whether a "special global one" controlled by an admin, or a "user controlled/user specific one").  Where it gets the content from is immaterial to the CSI driver and will be determined entirely by the PVC that is created and the config params it includes.

where the global vs user specific concerns will manifest is in the admission controller behavior:  how flexible the pod annotation is in terms of specifying where the content comes from, whether we auto-inject the PVC/volumemount for the global content always, etc.

edit: reading your comments below on namespace support, i may be operating under a misconception of how this gets wired, but let's chat on slack and you can educate me.  It seems like the PVC is going to have to at least have metadata that indicates what secret it wants.  Also specifying what namespace that secret comes from doesn't seem like a big addition, but it sounds like there may be complexities on the backend PV creation or secret watching.

Author: Yeah I also have an open question about if/how PVC metadata might get propagated to the CSI spec'ed methods/types.

But sure we can chat via slack / bluejeans during offices hours as needed.

Author: @bparees I believe https://kubernetes-csi.github.io/docs/volume-datasources.html holds the answer for how we can propagate the namespace information

A PVC datasource translates to 

```
type TypedLocalObjectReference struct {
	// APIGroup is the group for the resource being referenced.
	// If APIGroup is not specified, the specified Kind must be in the core API group.
	// For any other third-party types, APIGroup is required.
	// +optional
	APIGroup *string `json:"apiGroup" protobuf:"bytes,1,opt,name=apiGroup"`
	// Kind is the type of resource being referenced
	Kind string `json:"kind" protobuf:"bytes,2,opt,name=kind"`
	// Name is the name of resource being referenced
	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
}
```

We should be able to specify the namespace with that.

When I cycle off bugs and back to active proposal mgmt, I'll update the proposal with this and strike out the open question pertaining to this.

Author: Fair enough / good point.

As per namespace is out of scope, I'm going to
a) claim more than due diligence done with it, in at least exploring options through the CSI flows
b) remove this section entirely
c) leave the conversation thread unresolved so that it is easier for folks to discover what transpired and where we landed.

## Proposal

### User Stories [optional]

1. As a cluster admin with users who needs to have their Pod access a finite set of content on a host/node in read only 

> fashion with marking their Pod privileged, I want to provide my users a means of doing that.

Reviewer 2: ```suggestion
fashion without marking their Pod privileged, I want to provide my users a means of doing that.
```

> 2. As a user who needs to have their Pod access a finite set of content on a host/node in read only fashion with marking

Reviewer 2: ```suggestion
2. As a user who needs to have their Pod access a finite set of content on a host/node in read only fashion without marking
```

 - the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md) details the entire architecture,
 concepts, and protocol
 - so Volume provisioning could access the node/host filesystem, in particular the directory (and only the directory)
 where the secret contents will be copied to
 - As part of Volume creation/provisioning a key element is this: CSI has two step mounting process. staging and publishing.
 - They correspond to the endpoint's `NodeStageVolume` and `NodePublishVolume` implementations 
 - Staging is node local and publish is pod local. So we will probably want to stage the content once and publish it for each pod 
 (which is nothing but creating a bind mount for original volume)
 - That means we do not have to necessarily read and copy the secret(s) contents on every pod creation
 - the [CSI spec volume lifecycle section](https://github.com/container-storage-interface/spec/blob/master/spec.md#volume-lifecycle) 
 talks about a pre-provisioned volume.
 - If this is the right choice, perhaps this will help with performance, and how often we copy the Secret(s) contents
 onto the host/node
 

>  #### How can we boostrap this CSI driver implementation

Reviewer 2: ```suggestion
 #### How can we bootstrap this CSI driver implementation
```

---
title: inject-cross-namespace-secrets-onto-openshift-hosts-for-general-pod-consumption

authors:
  - "@gabemontero"

reviewers:
  - "@bparees"
  - "@adambkaplan"
  - "@derekwaynecarr"
  - "@gnufied"
  
approvers:
  - "@bparees"
  - "@adambkaplan"
  - "@derekwaynecarr"
  - "@gnufied"
  
creation-date: 2020-03-17

last-updated: 2020-03-18

<!-- status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced -->
status: provisional

  
---

> # Inject cross namespace secrets onto OpenShift Hosts for general Pod consumption

Author: Thanks for the suggestion @adambkaplan 

Though in cross referencing with other comments, like @bparees 's https://github.com/openshift/enhancements/pull/250#discussion_r394767547

`arguably this proposal(specific to the CSI driver) doesn't concern itself with whether the content is "global" or user namespaced.`

I'm inclined to *NOT* references namesapces in the title, and not use your suggestions as is.

Names and titles can be difficult, for sure.

My counter proposals:
-  `Expose Secrets/ConfigMap data via a CSI Driver to Pods`
-  `Share Secrets/ConfigMap data via a CSI Driver to Pods`
- `Expose Secrets/ConfigMaps via a CSI Driver to Pods`
- `Share Secrets/ConfigMaps via a CSI Driver to Pods`

> # Inject cross namespace secrets onto OpenShift Hosts for general Pod consumption

Reviewer 2: @gabemontero my point was that there's no actual concept of a "global" secret or a "namespaced" secret.  There are only secrets, they always live in namespaces, and if you want to mount one(that does not live in your pod's namespace) you will need this CSI driver.  Whether a secret is "global" or not is an artificial concept that will be determined by the admission control logic (e.g. treating all secrets that live in a particular special namespace as being "global" and thus always eligible for CSI driver copying, or treating all secrets annotated in a certain way as "global").

So referring to this CSI driver as allowing one to share a secret across namespaces is still reasonable, because even the "global" secrets are still, in reality, namespaced secrets that are being shared across namespaces (i.e. shared to pods that live in a different namespace).

> # Inject cross namespace secrets onto OpenShift Hosts for general Pod consumption

Author: Ok thanks for the clarification @bparees 

So with that in mind, on possible titles, I currently see 
- yours in https://github.com/openshift/enhancements/pull/250#issuecomment-602845720 .. `CSI driver to enable pods to mount cross-namespace secrets`
- we've got @adambkaplan 's `Share Secrets And ConfigMaps Across Namespaces via a CSI Driver` earlier in this thread
- and we got my 4 here in this thread

Care to pick which one you want us to settle with ?

> # Inject cross namespace secrets onto OpenShift Hosts for general Pod consumption

Reviewer 2: @adambkaplan's works for me.

> # Inject cross namespace secrets onto OpenShift Hosts for general Pod consumption

Author: part of next push

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [/] Design details are appropriately documented from clear requirements
- [/] Test plan is defined
- [/] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Open Questions [optional]

> 1. Should we manipulate our CSI plugin via the API from the CSI Spec outside the context of K8S Pod provisioning.  For 
>example, as part of a Operator/Controller co-located with the CSI plugin, can/should we call `DestroyVolme`, `NodeUnstageVolume`, or
>`NodeUnpublishVolume` when that Operator/Controller detects a change in any watched Secret resources, as part of updating
>the contents on the host/node?  Or should that Operator/Controller, as part of being hosted as a privileged `DaemonSet`, directly 
>manipulate the predefined location for the host/node content mapped to the CSI driver?  Sense is the latter is fine,

> >but confirmation from CSI SMEs would be good.

Author: thanks for the confirmation @jsafrane ... removing the open question and refs to it below

## Summary

Customer feedback on V4.x of OpenShift has included the desire to seamlessly inject certain content (usually of a sensitive nature)

> onto each host/node of a cluster for inclusion in the customer's pod.

Author: Moving up of "lead" part of next push.

I did not add anything about Entitlements being a concrete use case.  Readers can follow the ref to the parent PR to learn of that and other "broader proposal" scenarios.

> onto each host/node of a cluster for inclusion in the customer's pod.

Author: And per the other discussion thread, entitlements are now cited in the user stories.  Thanks.

At this time, the only means of doing so are 

- Declare `HostPath` based PV/PVCs on the Pod spec, on a per Pod basis, by each author of the Pod
- Leverage the 4.x machine configuration operator or MCO to update the host file system

There are several drawbacks with these approaches:

- `HostPath` volumes require the privileged bit being set on the Pod, which grants the Pod and Pod author overall permissions
that cluster administrators will not want, instead of access to just the specific piece of content on the host file system.
- Host injection via the MCO requires a node reboot for the updates to take effect.  That is generally deemed 
undesirable for this feature.  
- The MCO API has not been intuitive for the type of users to date who are interested in introducing and then consuming

> small collections of files on a host.

Reviewer 3: Most of this section above should be moved to the Alternatives section.

Author: part of next push

This proposal will describe a [CSI driver](https://github.com/container-storage-interface/spec/blob/master/spec.md) based

> means of OpenShift cluster administrators the means of injecting K8S Secrets onto each host/node in an OpenShift cluster, 

Reviewer 3: IMO adding this data to the node is an (important) implementation detail.

Author: Agreed.  Not sure if this was a plus one of some sort, or you were asking for a text change.  If the latter, please clarify @adambkaplan

and then allow owners/authors/users of Pods to opt into the consumption of the content of those Secrets during their 
Pod's execution.  The means of opting in will entail setting an annotation or label on an object vs. explicitly
defining the volume related API objects or volume plug points into the Pod.

This proposal was broken off from [a proposal on injecting RHEL entitlements/subscriptions](https://github.com/openshift/enhancements/pull/214)
**EDITORIAL NOTE:** when PR 214 merges the reference link will be adjusted to point to the merged document hosted on
https://github.com/openshift/enhancements.

## Motivation

Simplify overall usability for injecting and consuming Secret content onto the OpenShift hosts/nodes for use in user

> Pods.

Reviewer 2: this document is intended to include enough specifics that someone understands how it will be built/implemented (and can poke holes in that).  So it seems reasonable to have implementation details defined here.  (Whether those details are correct or not, with respect to the right way to implement this CSI driver, is what we are looking for feedback on)

> Pods.

Author: The motivation will be changed to `Simplify consumption of Secrets and ConfigMaps defined in one namespace in additional namespaces.` in the next push

### Goals

Provide alternatives to users specifying privileged Pods with HostPath volume.

> Provide alternatives to cluster administrators for using the MCO for injecting content onto nodes.

Reviewer 2: it's tied to the larger goal.  The larger goal (the reason we're going to build this CSI driver) is to share global secrets.  So one alternative to sharing global secrets via a CSI driver is to instead use the MCO to put the secret on the node and let pods use hostpath volumes to mount them.

You make a fair point that perhaps including that discussion here is confusing the issue.

> Provide alternatives to cluster administrators for using the MCO for injecting content onto nodes.

Reviewer 3: I agree here - the goal is to share Secrets across namespaces. ConfigMaps should also be included - in the subscription/entitlement use case we'll need to share both.

> Provide alternatives to cluster administrators for using the MCO for injecting content onto nodes.

Author: The Goals section will be reduced to `Provide easy to use sharing of `Secrets` and `ConfigMaps` defined in one namespace in additional namespaces.` in the next push

for watching for updates to the K8S Secrets being stored on the host/node by the associated CSI Driver of this proposal.
So if the cluster administrator realized on their own they have provided bad data, and then corrects it, this proposal
will watch for, discover, and propagate the update to the host/node filesystem.

2. Protecting this content from being viewed/copied by users. If it's available for the Pod to use, the user who can 
create a Pod can also read / view / exfiltrate the Secret contents. It is technically impossible to avoid that.

3. 4.x customer feedback has also revealed a desire for namespace level granularity when injecting/consuming
K8S Secret information available on the underlying host/node.  While this proposal will articulate some thought on 
how to do this, ultimately a solution based on a given Secret utilized across multiple namespaces is deemed "meets
minimum" for this proposal, and per namespace secret injection is out of scope.   

## Proposal

### User Stories [optional]

1. As a cluster admin with users who needs to have their Pod access a finite set of content on a host/node in read only 

> fashion without marking their Pod privileged, I want to provide my users a means of doing that.

Reviewer 3: I agree - these user stories are a bit abstract. IMO concrete use cases can help clarify this proposal, such as:

As a cluster admin with an OpenShift subscription
I want to share my entitlement keys and configuration with my workload pods
So that containers in my pods can install RHEL subscription content

> fashion without marking their Pod privileged, I want to provide my users a means of doing that.

Author: So my perspective on preferences around which details to put in which proposals, similar to what I noted in https://github.com/openshift/enhancements/pull/250#discussion_r397358585 , surfaces here as well again from my perspective.

And granted, some nuances of reviewer's suggestions may have become lost on me with the high, iterative churn.

So to be clear, and I think I agree if this is where you are coming from, for `User Stories` @adambkaplan @bparees :
- we are comfortable with specifics like the entitlement use case, even though it will be addressed in a broader fashion elsewhere?
- is this proposal allowed to only list a subset of those use cases, or does it need to be kept up to date as things evolve
- Or do we just want to put a ref to other proposal(s) where the scenarios are articluated ?

thanks

> fashion without marking their Pod privileged, I want to provide my users a means of doing that.

Reviewer 2: > we are comfortable with specifics like the entitlement use case, even though it will be addressed in a broader fashion elsewhere?

i'm comfortable with entitlements as a specific example of a use case.  The details of the proposal itself should not be tied to entitlements, they should just be sufficient to see that the proposal, when implemented, will address that use case (among others)

Another specific example of a use case worth including is distribution of CAs provided by the admin, to workloads.

> is this proposal allowed to only list a subset of those use cases, or does it need to be kept up to date as things evolve

you don't have to list all the ways people might use it, just enough to motivate why we need to do this.

 reading.
 - k8s provides Kubernetes CSI Sidecar Containers: a set of standard containers that aim to simplify the development and 
 deployment of CSI Drivers on Kubernetes.
 - There are CSI volume and persistent volume types in k8s.
 - per [k8s volume docs](https://kubernetes.io/docs/concepts/storage/volumes/#csi) "the CSI types do not support direct
 reference from a Pod and may only be referenced in a Pod via a `PersistentVolumeClaim` object" (direct quote)
 - However, @gnufied tells us that documentation is out of date and incorrect.  To quote him:  That is where [inline 
 CSI volumes come](https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html)
 - To that end, one of the sidecar containers, the [external provisioner](https://kubernetes-csi.github.io/docs/external-provisioner.html),
 helps with this
 - It watches Kubernetes `PersistentVolumeClaim` objects, and if the PVC 
 references a Kubernetes StorageClass, and the name in the provisioner field of the storage class matches the name 
 returned by the specified CSI endpoint `GetPluginInfo` call, it calls the CSI endpoint `CreateVolume` method to provision
 the new volume.

>  - there are also lifecycle methods for destroy, resize, etc.

Reviewer 3: > The volume can be probably injected by an admission plugin easily, I am not sure how to inject volumeMount, especially the mountPath.

We plan on having a separate admission controller that injects the volumeMount with the correct mountPath. This is out of scope for the proposal here.

>  - there are also lifecycle methods for destroy, resize, etc.

Reviewer 3: > The main benefit of ephemeral CSI volume is that you don't need a PVC in user's namespace (who's going to delete it and when?) and you don't need to implement Create/DeleteVolume and Stage/UnstageVolume in your driver.

This is a _huge_ benefit IMO. Is the goal for this feature to be GA by 4.6? Or alternatively, would we be comfortable shipping a CSI driver like this in 4.6 that utilizes the current beta API?

>  - there are also lifecycle methods for destroy, resize, etc.

Reviewer 4: > would we be comfortable shipping a CSI driver like this in 4.6 that utilizes the current beta API?

Yes, if the CSI driver is marked as Technical Preview. Which I would suggest for a first public release of any CSI driver anyway.

>  - there are also lifecycle methods for destroy, resize, etc.

Author: > I would strongly recommend using ephemeral CSI volumes, https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html

Yep @gnufied mentioned this briefly as well last week in https://github.com/openshift/enhancements/pull/250#discussion_r394562356 

> 
> It will be officially technical preview in OCP 4.5, but it's already beta in Kubernetes 1.17 and thus enabled by default in OCP 4.4. It's only not documented in OCP docs.
> 
> In that sense:
> 
> ```
> apiVersion: v1
> kind: Pod
> metadata:
>   name: some-pod
> spec:
>   containers:
>     ...
>      volumeMounts:
>         - name: entitlement
>           mountPath: /etc/entitlement
>   volumes:
>       - name: entitlement
>         csi:
>           driver: secretprovider.openshift.io
>           volumeAttributes:
>               namespace: openshift-config
>               name: entitlement
> ```
> 
> The `volume` can be probably injected by an admission plugin easily, I am not sure how to inject `volumeMount`, especially the `mountPath`.
> 
> The CSI driver then gets only `NodePublish` and `NodeUnpublish` call for each Pod that uses the volume. `NodePublish` does whatever you need - it could mount tmpfs and put the Secret data there and refresh it when it changes on API server. Or it could bind-mount a driver's directory (e.g. `/var/lib/secretprovider.openshift.io/<namespace>/<name>`) with the data, so you need to refresh it at one place for all pods on the node. With bind-mount, you must ensure that any Pod that needs the data can actually read it - for example have proper SELinux context and fs permissions. Read-only bind mount is strongly suggested.
> 
> The main benefit of ephemeral CSI volume is that you don't need a PVC in user's namespace (who's going to delete it and when?) and you don't need to implement Create/DeleteVolume and Stage/UnstageVolume in your driver.

Very very helpful drill down ... thanks

> 
> I'd recommend you to look at https://github.com/kubernetes-sigs/secrets-store-csi-driver - it implements a Secrets-like CSI driver that gets the secret data via plain `exec()`. It does not seem to implement refreshing of the data though.

Interestingly, that example was referenced in an initial conversation between @derekwaynecarr @bparees @gnufied and myself a few weeks ago but was subsequently ruled out.

Qutotes from @gnufied 
- https://coreos.slack.com/archives/GU5A8GSDR/p1582749384007000
- https://coreos.slack.com/archives/GU5A8GSDR/p1582749408007300

Still mulling as I type, and I still need to do a proper read of https://kubernetes-csi.github.io/docs/ephemeral-local-volumes.html, but at first blush, this feels like needing to specify both approaches as possible alternatives, at least in the interm, with pros/cons of each.

>  - there are also lifecycle methods for destroy, resize, etc.

Author: So completed my latest dive into https://github.com/kubernetes-sigs/secrets-store-csi-driver to better understand it. 

I tried to stand it up end to end like I did with csi-hostpath using its set of try it out steps.  I was unsuccessful, as using the 2 providers (azure and hashicorp) they cite did not work in my 4.x cluster (hostpath permissions, can't create directories, etc.) that I'm guessing they do not hit on vanilla k8s.  Or they have bugs.

In the end, I don't think it matters.  With more iterations and all the new info in this thread, I can piece together a roadmap using this style CSI plugin.

1) https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/pkg/azure/provider.go was a useful enough example of a storage class provider

2) the csi code at https://github.com/openshift/origin/tree/master/vendor/k8s.io/kubernetes/pkg/volume/csi makes more sense to me now

3) including https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/volume/csi/csi_mounter.go#L105-L254 which handles `NodePublishVolume` for both volsrc's and pvs, and maps both the VolumeAttributes @jsafrane noted in his Pod example, as well as vol src same namespace secret refs (https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/staging/src/k8s.io/api/core/v1/types.go#L1736) and pv/pvc any namesapce secret refs https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/staging/src/k8s.io/api/core/v1/types.go#L1674-L1705

4) https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/volume/csi/csi_util.go#L44-L55 for converting corev1.Secrets to the map[string]string signatures for secrets you see in the CSI APIs

With all that, I'll start articulating an ephemeral CSI volume option in addition to the existing PVC based one.

As previously noted, not needing to create a PVC is a big plus for ephemeral volumes.
Multi namespace secret refs seem like it could be a plus for the `CSIPersistentVolumeSource` (in conjunction with PVC) option, but seems like @jsafrane 's idea around using volumeAttributes to pass that metadata is an acceptable workaround.

 
 #### Some high level thoughts on what we need to inject/consume content on every host/node in an OpenShift cluster
   

>  - the CSI plugin would be hosted as a `DaemonSet` on the k8s cluster, to get per node/host granularity

Reviewer 4: On security note: what Secrets can this DaemonSet access? It probably needs access to entitlements and similar cross-namespace Secrets, but it should not get access to all Secret in the cluster.

Reviewer 2: it will have access to all secrets in the cluster because the admin will be able to use it to inject any secret from any namespace into any pod in any namespace, not just entitlements.

The security will be provided via admission logic which constrains who can create/mount these volumes in their pods (outside the scope of this proposal which is just about the CSI driver itself)

Author: So from my perspective, when consuming comments on this and #214, reviewers' perspectives on when to reveal particulars from #214 here and when to reveal particulars from this proposal there has been a bit of a moving target.

With that as a background @bparees , should I try to give some sort of summary of the admission logic piece here to help address readers' concerns like what @jsafrane listed here?

Author: Or just say "see ... for how security will be handled" ?

Reviewer 2: > Or just say "see ... for how security will be handled" ?

this one.  I would just explicitly state that this CSI driver is not responsible for security, it enables god mode and admins will need to deploy it in conjunction with something that limits who can use it/how it can be used.

Author: that works for me @bparees thanks

and then allow owners/authors/users of Pods to opt into the consumption of the content of those Secrets during their 
Pod's execution.  The means of opting in will entail setting an annotation or label on an object vs. explicitly
defining the volume related API objects or volume plug points into the Pod.

This proposal will depend on a subsequent proposal that will define the admission controller / webhooks required to
inject into a Pod on Pod creation a PVC citing as the storage class the CSI driver/plugin (represented by this proposal), 
if it sees the annontation/label previously noted, and validates any other constraints it chooses to impose from a permissions perspective.

## Motivation

Simplify overall usability for injecting and consuming Secret content onto the OpenShift hosts/nodes for use in user
Pods.

### Goals

Provide alternatives to users specifying privileged Pods with HostPath volume.

Provide alternatives to cluster administrators for using the MCO for injecting content onto nodes. 

### Non-Goals

1. While watching for secret changes when maintaining the new CSI based volume, validation of Secret contents or 

> corrective action for invalid Secret content is not addressed here. 

Author: Yeah as I read this now, this is not what I intended here.  In fact you'll see I mention watch for secrets and updating volumes below.

This was motivated and my attempt to generically capture https://github.com/openshift/enhancements/pull/214#issuecomment-600099270

I'll make another go of it.

>simple golang `os.MkdirAll` for example when creating mount access based volumes.  Can we do that in all our supported
>scenarios?  We've also received suggestions around tmpfs or read only bind-mounts of a directory from the CSI driver.
>**This question most likely can be deferred until implementation.**
> 3. SELinux implications will need to be properly understood.  Minimally, a consuming Pod needs the correct SELinux 
>context and fs permissions with bind-mounts for example.
>**This question most likely can be deferred until implementation.**
> 4. The EP document in https://github.com/openshift/enhancements/pull/169 needs to be reconciled with this document in
>some form or fashion.

## Summary

This proposal will describe a [CSI driver](https://github.com/container-storage-interface/spec/blob/master/spec.md) based
means for OpenShift cluster administrators to share `Secrets` / `ConfigMaps` defined in one namespace with
other namesapces. 

Any owners/authors/users of Pods in those namespaces can opt into the consumption of the content of those shared `Secrets` / 
 `ConfigMaps` during their Pod's execution.
 

> The precise means of opting in consuming the shared content are outside the scope of this proposal.

Reviewer 2: ```suggestion
The precise means of opting into consuming the shared content are outside the scope of this proposal.
```

This proposal was broken off from [a proposal on injecting RHEL entitlements/subscriptions](https://github.com/openshift/enhancements/pull/214)
**EDITORIAL NOTE:** if/when PR 214 merges the reference link will be adjusted to point to the merged document hosted on
https://github.com/openshift/enhancements.  Or if that PR is replaced by another, this proposal will point to that.

## Motivation

> Simplify consumption of `Secrets` and `ConfigMaps` defined in one namespace in additional namespaces.

Reviewer 2: ```suggestion
Simplify consumption of `Secrets` and `ConfigMaps` defined in one namespace by pods in other namespaces.
```

Reviewer 2: The motivation is probably more like "provide the foundational pieces for building a global/shared secret/configmap feature such that user pods in one namespace can consume secrets/configmaps from another namespace, such as provided by a cluster-admin"

Reviewer 2: (where the rest of said feature, dependent on this work,  is designed/defined/described in the other EP)

Author: ok I tried to capture that in an update I'm about to push

### Goals

> Provide easy to use sharing of `Secrets` and `ConfigMaps` defined in one namespace in additional namespaces.

Reviewer 2: ```suggestion
Provide easy to use sharing of `Secrets` and `ConfigMaps` defined in one namespace by pods in other namespaces.
```

### Non-Goals

1. This proposal assumes the sensitive content encapsulated in the `Secret` or `ConfigMap` is valid and correct for the user's purposes.
No validation of the content for specific usage within a Pod will be done.  That said, this proposal does provide 
for watching for updates to those objects.
So if the cluster administrator realized on their own they have provided bad data, and then corrects it, this proposal
will watch for, discover, and propagate the update when new Pods consuming the content are provisioned.

2. On protecting this content from being viewed/copied by users: if it's available for the Pod to use, the user who can 
create a Pod that can utilize this feature can also read / view / exfiltrate the Secret contents. It is technically 
impossible to avoid that.

> 

Reviewer 2: one more non-goal:  Providing security constraints on who can create these volumes/use them in their pods to access content from another namespace.  Security restrictions are being added/addressed in the related work, not here.

Author: part of next push

driver confirmation](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/docs/deploy-1.17-and-later.md#confirm-hostpath-driver-works), where that
confirmation is augmented to consider:

- the multiple replicas of the `DaemonSet`
- create/update/delete of the K8S Secret(s) designated for host/node injection are properly reflected on the volume

Also, there is already some e2e's around the existing csi-hostpath example.  The ginkgo test focus  `[sig-storage] CSI Volumes [Driver: csi-hostpath]`
shows up in several of the base e2e suites.

Some links related to that testing:
- https://github.com/openshift/origin/blob/master/cmd/openshift-tests/csi.go
- https://github.com/openshift/origin/tree/master/test/extended/testdata/csi
- https://github.com/openshift/origin/tree/master/vendor/k8s.io/kubernetes/test/e2e/storage/external

### Graduation Criteria

This should be introduced directly as a GA feature when it is implemented, with open questions addressed, including 
how exactly such CSI driver is installed/delivered to an OCP cluster, and how exactly we want to allow users to

> consume the provided CSI volume.

Reviewer 2: We will probably not want to GA this until the related work is also complete.  Without the admission controller+config logic to restrict who can actually use these volumes in a pod, this CSI driver by itself is just a giant security hole.

Author: Yeah I was unclear on this based on prior dialogues

I agree and will clarify here.

---
title: share-secrets-and-configmaps-across-namespaces-via-a-csi-driver

authors:
  - "@gabemontero"

reviewers:
  - "@bparees"
  - "@adambkaplan"
  - "@derekwaynecarr"
  - "@gnufied"
  
approvers:
  - "@bparees"
  - "@adambkaplan"
  - "@derekwaynecarr"
  - "@gnufied"
  
creation-date: 2020-03-17

last-updated: 2020-03-30

> 

Reviewer 5: fill in replaces with a link to #169 please

Author: part of next push

- [/] Test plan is defined
- [/] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Open Questions [optional]

> 1. Is a CSI `SecretProviderClass` based implementation recommended?  Received some varying guidance.  Also, per PR
>review from @jsafrane, of which I can only concur based on my examination, the example at https://github.com/kubernetes-sigs/secrets-store-csi-driver
>does not appear to handle refreshing of the data after initial provisioning.  So it does not appear to be something we
>can use as is.  But do we want to take on adding such a feature upstream to that repo? **This question most
>likely can be deferred until implementation.**
> 2. What will be the precise means of interacting with the host filesystem.  The hostpath plugin actually uses the 
>simple golang `os.MkdirAll` for example when creating mount access based volumes.  Can we do that in all our supported
>scenarios?  We've also received suggestions around tmpfs or read only bind-mounts of a directory from the CSI driver.
>**This question most likely can be deferred until implementation.**

> > 3. SELinux implications will need to be properly understood.  Minimally, a consuming Pod needs the correct SELinux 

Reviewer 5: not sure I see how this can be deferred. You can get some ways down the road, but I don't think you can deliver without having selinux working with restricted SCC

Author: Yeah I was trying to distinguish between merging this PR vs. actual delivery of the feature in OCP.

I'll update to make sure that is clearer.

thanks

>context and fs permissions with bind-mounts for example.
>**This question most likely can be deferred until implementation.**
> 4. The EP document in https://github.com/openshift/enhancements/pull/169 needs to be reconciled with this document in
>some form or fashion.

## Summary

This proposal will describe a [CSI driver](https://github.com/container-storage-interface/spec/blob/master/spec.md) based
means for OpenShift cluster administrators to share `Secrets` / `ConfigMaps` defined in one namespace with
other namesapces. 

> Any owners/authors/users of Pods in those namespaces can opt into the consumption of the content of those shared `Secrets` / 

Reviewer 5: This is a little vague.  I don't think an average user should be able to say, "my namespace shares this cool secret/configmap".  We had namespace to namespace links in the first RBAC implementation and the referential integrity and discovery problems became too great to bear.  A single namespace makes sense and so does a list curated by cluster-admins, but arbitrary links seem bad.

Reviewer 2: i think the mechanics of this are outside the scope of this EP (CSI driver) since the intent is that the CSI driver won't have any RBAC enforcement, it'll copy+mount whatever it is told to.

but when it does come to RBAC enforcement, i think our plan is that there's an admission controller which is configured by the admin, and that determines which namespaces/secrets/configmaps can be mounted by pods in which other namespaces.  So yeah, single/central configuration location, controlled by the admin.  Not random users offering up their secrets to random namespaces.

Author: see my slippery slope / new dividing line between this EP and #214 below

Reviewer 5: > i think the mechanics of this are outside the scope of this EP (CSI driver) since the intent is that the CSI driver won't have any RBAC enforcement, it'll copy+mount whatever it is told to.

Yeah, I don't think that's a good premise. I think that a CSI driver needs to be aware of permission changes to prevent updates which should not be applied.

I think a goal of this should be a CSI driver that be generally programmed for use or building on top of.  If we aren't careful with API definitions now, we won't be able to do that without rebuilding what's here.

As I noted below, I don't think an API that is sympathetic to these needs is difficult.

Reviewer 2: @gabemontero seems like this probably needs to be tightened up based on our discussion, it's now more like "users can opt into consuming content that a cluster-admin has chosen to make available"

 `ConfigMaps` during their Pod's execution.
 

> The precise means of opting into consuming the shared content are outside the scope of this proposal.

Reviewer 5: this concerns me.  Part of the problem is distributing the set of secret and configmap names that can be consumed.

Reviewer 2: That is what https://github.com/openshift/enhancements/pull/214 is for (that EP needs a retitling, i think.  That EP is intended to cover how we control which pods can consume which secrets, and auto-configuring pods(adding the volume) to do so in certain cases.

This EP is really just to implement the dependent bit of work of creating a CSI driver that allows any pod to mount any secret/configmap.  It is not about which pods are allowed to do it, or automation around configuring pods w/ certain secrets/configmaps.

Author: yeah there were pros and cons with splitting out the CSI driver bits from #214 as we decided to do

and a con is for folks coming in mid stream like @deads2k is here don't have the history

don't know if there are any guidelines upstream wrt how separate EPs are allowed to coordinate

it is a slippery slope, but based on @deads2k `ProjectedResource` suggestion below, I'm thinking at least part of the opt in piece neeeds to be here in this EP

I'll take a stab at it, change this line to reflect the new dividing line between this EP and #214, 
and we can then see what form of consensus we have

This proposal was broken off from [a proposal on injecting RHEL entitlements/subscriptions](https://github.com/openshift/enhancements/pull/214)
**EDITORIAL NOTE:** if/when PR 214 merges the reference link will be adjusted to point to the merged document hosted on
https://github.com/openshift/enhancements.  Or if that PR is replaced by another, this proposal will point to that.

## Motivation

Help simplify consumption of `Secrets` and `ConfigMaps` defined in one namespace by pods in other namespaces by providing
the foundational pieces for building a shared secret/configmap feature such that user pods in one namespace can consume
secrets/configmaps from another namespace (which is presumably maintained by a cluster admin).

> ### Goals

Reviewer 5: should have a goal of keeping the content up to date.

Reviewer 2: as in "when the configmap changes, the container filesystem content needs to reflect those changes"?

If so yes, i think Gabe has that mentioned elsewhere but making it an explicit goal makes sense to me.

Author: part of next push

Provide easy to use sharing of `Secrets` and `ConfigMaps` defined in one namespace by pods in other namespaces.

### Non-Goals

1. This proposal assumes the sensitive content encapsulated in the `Secret` or `ConfigMap` is valid and correct for the user's purposes.
No validation of the content for specific usage within a Pod will be done.  That said, this proposal does provide 
for watching for updates to those objects.
So if the cluster administrator realized on their own they have provided bad data, and then corrects it, this proposal
will watch for, discover, and propagate the update when new Pods consuming the content are provisioned.

2. On protecting this content from being viewed/copied by users: if it's available for the Pod to use, the user who can 
create a Pod that can utilize this feature can also read / view / exfiltrate the Secret contents. It is technically 
impossible to avoid that.

> 3. Providing security constraints on who can create these volumes and/or use them in their pods to access secrets/configmaps

Reviewer 5: Before such a feature can be considered complete, this needs to be addressed.  This should be a goal. You don't need to solve it for prototypes, but we cannot release software without some kind of restrictions on how content is exposed.  This type of problem dates back to the days of cross-namespace RBAC rolebinding refs. It was a big mistake to allow it.

Reviewer 2: absolutely, see the "GA requrements" at the end of this EP.  We will not ship this CSI driver w/o the additional admission logic that makes it safe to install.

Reviewer 2: (but to be clear, that does not make it a goal of this EP to address that RBAC requirement.  The parent EP i referenced has that goal+responsibility).

Author: will reference GA reqs below and clarify this is handled in separate EP

Reviewer 5: > (but to be clear, that does not make it a goal of this EP to address that RBAC requirement. The parent EP i referenced has that goal+responsibility).

I maintain that a prototype really needs to build this in from the start to be viable.  We have experienced many times in this project and the upstream project extreme difficulty in retconing security.  You can see below the impact this has in API design.  It's particularly difficult to consider now to shape the API and should be fairly easy to add to a prototype as a distinct unit.

from another namespace.  That will be addressed in a separate proposal as noted in the Summary. 

## Proposal

### User Stories [optional]

1. As a cluster admin with an OpenShift subscription, I want to share the entitlement keys and configuration with my 
users' workload Pods so that containers in those Pods can install RHEL subscription content.

2. As a cluster admin with a set of CAs pertinent to various users's workloads, I want to share those CAs with my 
users' workload Pods so that containers in those Pods can utilize those CAs when performing TLS based communication.

3. See [this proposal](https://github.com/openshift/enhancements/pull/214) for additional user scenarios that could 
utilize this proposal.

### Implementation Details/Notes/Constraints [optional]

#### Install

A comparison of installing the implementation of this proposal via OLM vs. CVO was done in another enhancement proposal.  

> The net result was to go with CVO.

Reviewer 5: This result surprises me.  Such a CSI plugin has immense value for the community and is a fairly open space for innovation.  I would hope to see this distributed via OLM first and by the CVO second (if at all).

Reviewer 2: definitely a point for on-going discussion, but absent a way to have OLM operators installed on day one, making this an OLM operator really hamstrings our ability to deliver the feature we are ultimately building this for, seamless consumption of entitlements.

Author: I'll re-add this as an open question

It had been an open question until I saw the exchange @deads2k between you and @jsafrane in https://github.com/openshift/enhancements/pull/139/files/1ff25b741829deac9e6ba3826797c31a2f3430df#r391616151 and https://github.com/openshift/enhancements/pull/139/files/1ff25b741829deac9e6ba3826797c31a2f3430df#r391761362

I'm taking your comment here that my interpretation that a consensus was reached was incorrect.

Reviewer 2: @gabemontero i think we probably need @derekwaynecarr to help us make a decision on this one.

Reviewer 5: How far off are we from being able to have another operator (or even the CVO) require an OLM operator to be present?  I think that needs to be an objective in order to migrate any operator to an OLM based one.

 - To that end, one of the sidecar containers, the [external provisioner](https://kubernetes-csi.github.io/docs/external-provisioner.html),
 helps with this
 - It watches Kubernetes `PersistentVolumeClaim` objects, and if the PVC 
 references a Kubernetes StorageClass, and the name in the provisioner field of the storage class matches the name 
 returned by the specified CSI endpoint `GetPluginInfo` call, it calls the CSI endpoint `CreateVolume` method to provision
 the new volume.
 - there are also lifecycle methods for destroy, resize, etc.
 - On the more recently introduced inline ephemeral mode of which @jsafrane has strongly suggested to us and which 
 @gnufied also noted ... 
 - In the ephemeral case, the Pod spec lists a volume and is considered to originate or "inline" it.
 - The `csi` subfield is used, and the storage provider's name is listed as with the PVC based approach.
 
 ```yaml
apiVersion: v1
kind: Pod
metadata:
  name: some-pod
spec:
  containers:
    ...
     volumeMounts:
        - name: entitlement
          mountPath: /etc/entitlement
  volumes:
      - name: entitlement
        csi:
          driver: secretprovider.openshift.io
          volumeAttributes:

>               namespace: openshift-config

Reviewer 5: I would not expect an API shaped like this.

Author: After going over this comment a few times (bouncing between a few interpretations), and cross referencing with your `ProjectedResource` suggestion below, I'm landing with this
- in order to hide the originating coordinates as you suggest below, a reference to the namespace the shared secret is in, and the shared secret name itself, should **NOT** be present in the k8s CSI volume definition in this example
- instead, a reference to `ProjectedResource` should exist in the Pod (presumably an annotation, but you didn't previously specify, so please advise further if needed)
- the admission plugin will decide based on the ACL checks if the pod's SA is even allowed to use said `ProjectResource`
- if so, it will mutate the pod to add the k8s CSI volume def, similar to this example, but with a notable exception
- namely, in that k8s CSI volume def, the volumeAttributes would list the requested `ProjectedResource`
- the CSI driver will look up the `ProjectedResrouce`, then look up the referenced Secret/Configmap, and go from there

Sound close @deads2k ?

Reviewer 5: ```
csi:
  driver: projectedresource.storage.openshift.io
    volumeAttributes:
      projectedResourceName: the-projected-resource
```

this way the information provided matches the projectedresource (clusterscoped) and cannot be used for user-defined fishing of namespaces.

means for OpenShift cluster administrators to share `Secrets` / `ConfigMaps` defined in one namespace with
other namespaces. 

Any owners/authors/users of Pods in those namespaces can opt into the consumption of the content of those shared `Secrets` / 
 `ConfigMaps` during their Pod's execution.

This proposal was broken off from [a proposal on injecting RHEL entitlements/subscriptions](https://github.com/openshift/enhancements/pull/214)
**EDITORIAL NOTE:** if/when PR 214 merges the reference link will be adjusted to point to the merged document hosted on
https://github.com/openshift/enhancements.  Or if that PR is replaced by another proposal, or this proposal, this proposal will 
point that out.

## Motivation

Simplify provision and consumption of `Secrets` and `ConfigMaps` defined in one namespace by pods in other namespaces. 

### Goals

- Provide easy to use sharing of `Secrets` and `ConfigMaps` defined in one namespace by pods in other namespaces.

- When a shared `Secret` or `ConfigMap` changes, the container filesystem content the CSI driver injects into the Pod
needs to reflect those changes.

> - When a cluster-admin revokes access to shared `Secrets` or `ConfigMaps` that as previously granted, that data should

Author: thanks part of next push

for watching for updates to those objects.
So if the cluster administrator realized on their own they have provided bad data, and then corrects it, this proposal
will watch for, discover, and propagate the update when new Pods consuming the content are provisioned.

2. Protecting this content from being viewed/copied by users: if it's available for the Pod to use, the user who can 
create a Pod that can utilize this feature can also read / view / exfiltrate the Secret/ConfigMap contents. It is technically 
impossible to avoid that.

3. This proposal for now assumes linux only for the underlying host/node.  No windows, etc. 

## Proposal

### User Stories [optional]

1. As a cluster admin with an OpenShift subscription, I want to share the entitlement keys and configuration with my 
users' workload Pods so that containers in those Pods can install RHEL subscription content.

2. As a cluster admin with a set of CAs pertinent to various users's workloads, I want to share those CAs with my 
users' workload Pods so that containers in those Pods can utilize those CAs when performing TLS based communication.

> 3. As a cluster admin whose user workloads include kmod based scenarios, I want to share the necessary information to
> facilitate those scenarios across multiple namespaces.

Reviewer 3: Suggestion for cleaning this up a bit:
```suggestion
3. As a cluster admin whose user workloads install kernel modules on nodes, I want to share the necessary information to
facilitate installation of kernel modules across multiple namespaces.
```

Author: thanks part of next push

4. As a cluster admin who has to rotate credentials shared via this proposal, I want to be able to remove access to 

> stale versions of those credentials. 

Reviewer 2: You want to remove access to new versions of the credentials(i.e. stop updating content within the consuming pod), not stale versions (the stale versions may have already been copied by the user in question anyway, so whether we delete that content out of the pod, or leave it in place, doesn't particularly matter from a security perspective.  Whether it's nicer to remove it or leave it, from the perspective of a user running the pod, open to debate i suppose).

Author: OK reconciling with https://github.com/openshift/enhancements/pull/250#discussion_r407620058

```
ask storage if removal is more normal than leaving stale content. I'm ok with either.

Describe the scenario we talked about with a cluster-admin trying to revoke access and rotate creds. That explains why it is important to make the SAR check.
```

Yeah, in hindsight I perhaps conflated those two points when I should not have.

Will update thanks.

described in that document (namely OLM).  The recommendation from the [storage SMEs was to go with CVO](https://github.com/openshift/enhancements/pull/139/files#r391761362)
was to not use OLM, and use CVO instead.
- Using CVO and thus making it a day 1 install operation meets the ease of use targets we are trying to
reach with this proposal
- Crafting OLM dependencies between this proposal and [related proposals](https://github.com/openshift/enhancements/pull/214)
as part of an OLM based install at the time of this writing is immature, at least for some reviewers.
- However, other reviewers wonder if we are in fact close from having either OLM or CVO operators being able to require
an OLM operator to be present.
- Using OLM avoids further increasing the install payload size.
- Using OLM avoids unexpected hiccups in the feature's installation impacting initial cluster install or cluster upgrade.

#### End to end flow (from creation, to opt-in, to validation, and then injections)

An admin creates a new cluster level custom resource for encapsulating the sharable `Secret` or `ConfigMap`.

```yaml
type: ProjectedResource
apiVersion: projected-resource.storage.openshift.io/v1

> name: SharedExternalAccessSecret

Reviewer 3: Do you mean something like the following?
```suggestion
metadata:
  name: shared-cool-secret
```

Author: yeah this was a copy/paste from a manual suggestion, without full yaml validation by me... I'll add these additions thanks

spec:
  backingResource:

>     type: Secret

Reviewer 3: Shouldn't we support the standard `kind`/`apiVersion` fields? IMO this future-proofs the API for other types of backing resources.

Author: sure ... again, not fully linted k8s yaml ... will update

    namespace: ns-one
    name: cool-secret
  description: "used to access my internal git repo, RHEL entitlements, or something else cool"
status:

>   conditions: available

Reviewer 2: what do these conditions reflect?  whether or not the backing resource actually exists?

Author: to be honest 
1) I left David's original suggestion as is ... I think he was alluding to the fact we would want status conditions at some point
2) first blush, I'm not sure I'm inclined to get too detailed at this time on what the conditions will be
3) I think prototyping and learning how some of the OS level filesystem mechanics flesh out might have bearing here
4) as such, at this time, I'm more inclined to just say
```
status:
  condtions:
 ...
```

and not get specific at all.

Thoughts?

Reviewer 2: i'm just looking for an understanding of the intended general nature of the conditions, not the specific values.

since this resource exists independent of any pods/volumes mounting it, I would not expect the conditions to reflect anything to do w/ the CSI driver/OS filesystem mechanics.

Author: yeah I don't know ... I could see the status reflecting which pods are using the secret/configmap encapsulated by this new CRD.  If we went with that, that update be initiated at the CSI driver level (after the SELinux concerns are reconciled) vs. the admission plugin.

Or it could be something entirely different.

Given your resolution of this thread, I'm going to go with `...` and we'll drill down on this once we get to implementation.

thanks

```

> And sets up ACL/RBAC to control who can list/get such objects.

Reviewer 3: May be nice to provide an example RBAC that provides read access:
```
rules:
- apiGroups:
  - projected-resource.storage.openshift.io
  resources:
  - projectedResources
  resourceNames:
  - shared-cool-secret
  verbs:
  - get
  - list
  - watch
```

Author: added thanks

  - "@gnufied"
  - "@jsafrane"
  - "@deads2k"
  
approvers:
  - "@bparees"
  - "@adambkaplan"
  - "@derekwaynecarr"
  - "@deads2k"
 
replaces:
  - "https://github.com/openshift/enhancements/pull/169"
  
creation-date: 2020-03-17

last-updated: 2020-04-14

<!-- status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced -->
status: provisional

  
---

# Share Secrets And ConfigMaps Across Namespaces via a CSI Driver

## Release Signoff Checklist

- [/] Enhancement is `implementable`
- [/] Design details are appropriately documented from clear requirements
- [/] Test plan is defined
- [/] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Open Questions [optional]

> 1. If data on the CSI volume represents data the Pod's SA no longer has access to that data because of ACL/RBAC changes,
>from a storage/CSI volume perspective are there any best practices on how to deal with the data?  Can we just remove (via
>an unmount of the assocated tmpfs) the data on the file system via a `NodeUnpublishVolume` call?  Do we have to leave 

> >the data around for existing pods?

Author: My question was entirely from uncertainty and what a CSI driver might expect as part of removing access to data.

Your statement here answers it for me as "no", so I'll remove the question.

I'll also add your point on kill pod / for security sake / etc.

thanks

> > 2. CVO vs. OLM for delivery.  May need @derekwaynecarr to cast the deciding vote.  Latest pro-con list below.

Author: I'll add that nuance to the question

 - so each pod gets its own tmpfs
 - the kubelet leverages tmpfs in a similar fashion already

#### End to end flow (from creation, to opt-in, to validation, and then injections)

An admin creates a new cluster level custom resource for encapsulating the sharable `Secret` or `ConfigMap`.

```yaml
kind: ProjectedResource
apiVersion: projected-resource.storage.openshift.io/v1
metadata:
  name: shared-cool-secret
spec:
  backingResource:
    kind: Secret
    apiVersion: v1
    namespace: ns-one
    name: cool-secret
  description: "used to access my internal git repo, RHEL entitlements, or something else cool"
status:
  conditions:
  ...
```

And sets up ACL/RBAC to control who can list/get such objects.

```yaml
rules:
- apiGroups:
  - projected-resource.storage.openshift.io
  resources:
  - projectedResources
  resourceNames:
  - shared-cool-secret
  verbs:
  - get
  - list
  - watch

```

A user queries this new type to see what is available to her/him.

Then, to opt in, a user created Pod can cite a CSI volume references (see below for a more detailed explanation of them) that 
references our new type:

```yaml

> csi:

Reviewer 4: IMO, whole "tree" would be nice. `csi` itself does not say where it belongs.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: some-pod
spec:
  containers:
    ...
  volumes:
      - name: secret
        csi:
          driver: projectedresources.storage.openshift.io
          volumeAttributes:
            projectedResourceName: shared-cool-secret
```

Author: I have the Pod level yaml a little bit further down but sure I can replicate that from the get go

    volumeAttributes:
      projectedResourceName: the-projected-resource

```

For pre-existing objects that encapsulate a Pod, a new, documented annotation can be declared to facilitate injection
of the CSI driver volume specification 

A new admission plugin introduced as part of this feature can then:
- mutate Pods with this annotation, adding the CSI driver volume specification, if the Pod `ServiceAccount` is allowed sufficient 
access the `ProjectedResource` (via a SAR check)
- if a Pod already happens to have the CSI Volume specified with reference to a `ProjectedResource`, again see if the 
associated `ServiceAccount` for the Pod has sufficient access to the `ProjectedResource`
- if sufficient access does not exist, the admission plugin marks the Pod as invalid/forbidden.
- if sufficient access does exists, the `volumeAttributes` are also updated to add a key/value pair for the requesting
`ServiceAccont` (including both its name and namespace)

```yaml
csi:
  driver: projectedresources.storage.openshift.io
    volumeAttributes:
      projectedResourceName: the-projected-resource
      requestingNamespace: the-users-namespace

>       requestingServiceAccount: the-sa-the-user-is-using-for-this-object

Reviewer 4: Why is namespace / service account necessary? Admission plugin can see it in the Pod spec it is admitting and CSI driver can get it in `NodePublishVolume` call.

Author: it was the ability of the "CSI driver can get it in `NodePublishVolume` call" that I had not discovered to date

on your prompting, I revisited secrets-store-csi-driver and sure enough I see 

```
		parameters["csi.storage.k8s.io/pod.name"] = attrib["csi.storage.k8s.io/pod.name"]
		parameters["csi.storage.k8s.io/pod.namespace"] = attrib["csi.storage.k8s.io/pod.namespace"]
```

where `attrib` comes from `NodePublishVolumeRequest.GetVolumeContext()`

I'm also finding those strings in vendored k8s code, as well as `csi.storage.k8s.io/serviceAccount.name`

I'll update accordingly, thanks

Reviewer 4: Yes, that's how the CSI driver gets it. `CSIDriver.spec.podInfoOnMount: true` must be set to enable this behavior on Kubernetes side. See a sample CSIDriver deployment : https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/deploy/kubernetes-1.17/hostpath/csi-hostpath-driverinfo.yaml#L12

Author: cool ... adding that note

```

The CSI Driver, aside from all the CSI and filesystem bits (see below for details), performs:
- reads of the `ProjectedResource` object referenced in the `volumeMetadata`
- reads of the `Secret` or `ConfigMap` noted in the `ProjectedResource`, storing it on disk on initial
reference to said `Secret` or `ConfigMap`
- creation of a tmpfs filesystem for each requesting `ServiceAcount` that contains the `Secret` / `ConfigMap` data
- creation of k8s watches `ProjectedResources`, and for `Secret` / `ConfigMaps` in the namespaces listed in each discovered
`ProjectedResources`
- final returning of a volume that is based off of the `ServiceAccount` specific link for the data noted in the 
`volumeAttributes`

After initial set up of things, steady state CSI driver activities include:
- as updates of the watched `Secrets` and `ConfigMaps` come in, SAR checks will be done for each of the previous requesting 
`ServiceAccounts` being tracked against the updated `Secret` and `ConfigMap`
- for any `ServiceAccounts` that no longer have access, their tmpfs filesystems are unmounted, effectively removing their access to 

> the data

Reviewer 4: Not sure this is going to work due to mount propagation. It's probably OK to leave it as it is now, but test this carefully.

Author: I'll add a comment that we need to keep an eye on this thanks

- and use of the k8s util mount API's
- it already employs `DaemonSets`
- the `SecretProviderClass` examples did not work out of the box on OpenShift, but we are not going down that path 
anyway.

### Risks and Mitigations

While this proposal settles on installing via CVO, that decision has ramifications whose discussion are noted above.
It is at least conceivable the need to revisit that decision may arise when implementation starts.

Getting the flow and various stages from the CSI specification understood and correctly utilized so that we minimize
the amount of times the Secret is read and stored on the local filesystem of each host/node is important.  We want
to minimize local file system writes to when the Secret actually changes, and not when we provision the volume to 
a new Pod. 

## Design Details

### Test Plan

Verification of this proposal should be containable around e2e's that automate various forms of [the basic Hostpath
driver confirmation](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/docs/deploy-1.17-and-later.md#confirm-hostpath-driver-works), where that
confirmation is augmented to consider:

- the multiple replicas of the `DaemonSet`

> - create/update/delete of the K8S Secret(s) designated for host/node injection are properly reflected on the volume

Reviewer 4: Please explicitly add a test for RBAC revocation (i.e. user looses access to ProjectedResource while running a pod that uses it)

Author: part of next push

described in that document (namely OLM).  The recommendation from the [storage SMEs was to go with CVO](https://github.com/openshift/enhancements/pull/139/files#r391761362)
was to not use OLM, and use CVO instead.
- Using CVO and thus making it a day 1 install operation meets the ease of use targets we are trying to
reach with this proposal
- Crafting OLM dependencies between this proposal and [related proposals](https://github.com/openshift/enhancements/pull/214)
as part of an OLM based install at the time of this writing is immature, at least for some reviewers.
- However, other reviewers wonder if we are in fact close from having either OLM or CVO operators being able to require
an OLM operator to be present.
- Using OLM avoids further increasing the install payload size.
- Using OLM avoids unexpected hiccups in the feature's installation impacting initial cluster install or cluster upgrade.

#### End to end flow (from creation, to opt-in, to validation, and then injections)

An admin creates a new cluster level custom resource for encapsulating the sharable `Secret` or `ConfigMap`.

```yaml

> type: ProjectedResource

Reviewer 5: specify the apiVersion here too.  `storage.openshift.io/v1` and negotiate with @openshift/storage for whether they want a sub-group on it.  I'm ok with a stutter to scope it properly, so `ProjectResource` in `projected-resource.storage.openshift.io` is fine with me.

Author: added in next push

Reviewer 3: I assume we mean `kind` here?

name: SharedExternalAccessSecret
spec:
  backingResource:
    type: Secret
    namespace: ns-one
    name: cool-secret
  description: "used to access my internal git repo, RHEL entitlements, or something else cool"
status:
  conditions: available
```

And sets up ACL/RBAC to control who can list/get such objects.

A user queries this new type to see what is available to her/him.

Then, to opt in, a user created Pod can cite a CSI volume references (see below for a more detailed explanation of them) that 
references our new type:

```yaml
csi:

>   driver: projectedresource.storage.openshift.io

Reviewer 5: nit: projectedresources I think

Author: part of next push

    volumeAttributes:
      projectedResourceName: the-projected-resource

```

> For pre-existing objects that encapsulate a Pod, a new, documented annotation can be declared to facilitate injection

Reviewer 5: if we expose names in a spot users can read, we may not need to have an annotation that does this.  The user can simply list the ProjectedResource in their pod because the name will be presumed stable and discoverable in a way the the previous: specify-a-namespace-and-name was not.

Author: I'm missing something here @deads2k  ... *how* exactly does the user list the `ProjectedResource` in their pod?  Or another way, what yaml snippet are you referring to when you say this?

thanks

Reviewer 2: wouldn't they be listing it in their pod volume definitions?

Reviewer 3: I agree we don't need an annotation for the CSI driver. For entitlements we will need to build on top of this:

1. An operator that does the following:
  a. Creates the `ProjectedResource` for the entitlement `Secrets` and `ConfigMaps` that live in `openshift-config`
  b. Creates RBAC rules for the entitlements.
  c. Adds appropriate `ClusterRoleBindings` (perhaps based on user config - will address in a future EP).
2. An admission webhook which does the following:
  a. Looks for the entitlement annotation on all created pods
  b. Adds the CSI volume mount(s) for the entitlement `ProjectedResources`, and adds appropriate `volumeMount` options for all containers.

Author: > wouldn't they be listing it in their pod volume definitions?

Yes, perhaps, @bparees 

I'm looking to confirm that interpretation is what @deads2k was alluding to.

Author: > I agree we don't need an annotation for the CSI driver. For entitlements we will need to build on top of this:
> 
>     1. An operator that does the following:
>        a. Creates the `ProjectedResource` for the entitlement `Secrets` and `ConfigMaps` that live in `openshift-config`
>        b. Creates RBAC rules for the entitlements.
>        c. Adds appropriate `ClusterRoleBindings` (perhaps based on user config - will address in a future EP).
> 
>     2. An admission webhook which does the following:
>        a. Looks for the entitlement annotation on all created pods

The annotation I'm referring to is this one @adambkaplan.  Per prior discussions I've moved the line a bit on which bits we discuss here vs. in say an admission plugin proposal.  It was becoming clear from prior reviews of this proposal was that some sort of high level end of end flow was needed in this doc.

That said, per earlier comments on this thread, and making sure I fully understand @deads2k comment, I may either remove this line, or much further clarify that it is a nice to have.  Probably will do the former and remove, but am not pulling the trigger just yet.

>        b. Adds the CSI volume mount(s) for the entitlement `ProjectedResources`, and adds appropriate `volumeMount` options for all containers.

Author: So @deads2k and I talked ... basically we are on the same page.  Nothing blocking what is currently here, but I'm going to add some clarifying text in my next push to cover the various bases touched on this thread.

Author: ok @bparees @deads2k @adambkaplan I've pushed an update to the end to end flow with those clarifying updates I referenced in https://github.com/openshift/enhancements/pull/250#discussion_r409133767 to touch about "to annotate or not to annotate" as well as some minor additional details here and there

    volumeAttributes:
      projectedResourceName: the-projected-resource
      requestingNamespace: the-users-namespace
      requestingServiceAccount: the-sa-the-user-is-using-for-this-object
```

The CSI Driver, aside from all the CSI and filesystem bits (see below for details), performs:
- reads of the `ProjectedResource` object referenced in the `volumeMetadata`
- reads of the `Secret` or `ConfigMap` noted in the `ProjectedResource`, storing it on disk on initial
reference to said `Secret` or `ConfigMap`
- creation of filesystem links for each requesting `ServiceAcount` to the actual files for the `Secret` / `ConfigMap` data
- creation of k8s watches `ProjectedResources`, and for `Secret` / `ConfigMaps` in the namespaces listed in each discovered
`ProjectedResources`
- final returning of a volume that is based off of the `ServiceAccount` specific link for the data noted in the 
`volumeAttributes`

After initial set up of things, steady state CSI driver activities include:
- as updates of the watched `Secrets` and `ConfigMaps` come in, SAR checks will be done for each of the previous requesting 
`ServiceAccounts` being tracked against the updated `Secret` and `ConfigMap`

> - for any `ServiceAccounts` that no longer have access, their filesystem links are removed, effectively removing their access to 

Reviewer 5: ask storage if removal is more normal than leaving stale content.  I'm ok with either.

Describe the scenario we talked about with a cluster-admin trying to revoke access and rotate creds.  That explains why it is important to make the SAR check.

Author: Adding the remove/leave stale data question in the open questions section as part of the next push.  Will follow up more directly with storage SMEs as feedback dictates.

Adding the scenario to the goals section in the next push, then referencing it here.

  containers:
    ...
     volumeMounts:
        - name: shared-item-name
          mountPath: /path/to/shared-item-data
  volumes:
      - name: entitlement
        csi:

>           driver: secretprovider.openshift.io

Reviewer 5: update for projectedresources and the name of the projectedresource?

Author: yep will make consistent with prior yam in doc in next push

- if a Pod already happens to have the CSI Volume specified with reference to a `ProjectedResource`, again the plugin will see if the 
associated `ServiceAccount` for the Pod has sufficient access to the `ProjectedResource`
- if sufficient access does not exist, the admission plugin marks the Pod as invalid/forbidden.
- if sufficient access does exists, the admission plugin adds a CSI volume into the Pod's `volume` array, to match the
example noted above.

The CSI Driver, aside from all the CSI and filesystem bits (see below for details), performs:
- reads of the `ProjectedResource` object referenced in the `volumeMetadata` which is available in the `NodePublishVolumeRequest`

> that comes into `NodePublshVolume`

Reviewer 4: Typo: `NodePublshVolume`

Author: pushing fix in a sec
